### PR TITLE
Add config resolve and config command line params

### PR DESCRIPTION
**Pull request by mandarjog**
This PR wires up config struct and uses config resolve to select and dispatch `config.Combined` to aspectManagers.

1. Remove proto.struct from everywhere, it is not longer needed
2. Up test coverage for files touched.
3. Removed repeated validation code from managers.
    The aspect managers are expected to get a validated config.
    It is ok to panic since we will ensure that the dispatcher will catch all the panics.
4. Add adapter.Inventory that maintains inventory of all adapters (builders) available in the system
5. Add aspect.Inventory that maintains inventory of all aspects (aspect Managers) in the system
6. Use (serviceConfig and globalConfig) as  command line parameters for config input.

This part will be replaced by a single URL that points to a repository location.

---
This change is [Reviewable](https://reviewable.io/reviews/istio/mixer/189)

### CODE REVIEW

File: pkg/config/api.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	// CheckMethod represents Check operation of the mixer api.
+	CheckMethod APIMethod = iota
+	// ReportMethod represents Report operation of the mixer api.
+	ReportMethod
+	// QuotaMethod represents Quota operation of the mixer api.
+	QuotaMethod
+)
+
+// APIMethods is the authoritative place that lists supported api methods.
+var APIMethods = []APIMethod{CheckMethod, ReportMethod, QuotaMethod}
```
Comment (geeknoid) [inline pkg/config/api.go]: Somewhat ikky to expose a mutable global variable outside of the package. Probably should be made internal and have a function expose the array.

File: pkg/config/manager_test.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+		if mt.errStr == "" && le == nil {
+			continue
+		}
+
+		if !strings.Contains(le.Error(), mt.errStr) {
+			t.Errorf("[%d] Unexpected error. Expected %s\nGot: %s\n", idx, mt.errStr, le)
+		}
+
+		time.Sleep(time.Second * 1)
```
Comment (geeknoid) [inline pkg/config/manager_test.go]: Why is this here? Can we have real synchronization? If not, then at least a comment explaining why this hack is here.

File: pkg/config/manager.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+)
+
+// ChangeListener listens for config change notifications.
+type ChangeListener interface {
+	ConfigChange(cfg *Runtime)
+}
+
+// Manager represents the config Manager.
+// It is responsible for fetching and receiving configuration changes.
+// It applied validated changes to the registered config change listeners.
```
Comment (geeknoid) [inline pkg/config/manager.go]: applied -> applies

File: pkg/config/manager.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+	scSha   [sha1.Size]byte
+	gcSha   [sha1.Size]byte
+
+	lock      *sync.RWMutex
+	lastError error
+}
+
+// ManagerArgs are constructor args for a manager
+type ManagerArgs struct {
+	Eval          expr.Evaluator
```
Comment (geeknoid) [inline pkg/config/manager.go]: All these fields deserve comments.

File: pkg/config/manager.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+	AspectF       ValidatorFinder
```
Comment (geeknoid) [inline pkg/config/manager.go]: Please expand "F" to "Finder"

File: pkg/config/manager.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+// api.Handler listens for config changes.
+type Manager struct {
+	ManagerArgs
+
+	cl      []ChangeListener
+	closing chan bool
+	scSha   [sha1.Size]byte
+	gcSha   [sha1.Size]byte
+
+	lock      *sync.RWMutex
```
Comment (geeknoid) [inline pkg/config/manager.go]: Why is this a pointer to a lock, instead of just an instance of a lock inline?

File: pkg/adapter/registrar.go [hunk 7]
```diff
@@ -17,18 +17,18 @@ package adapter
...
-	RegisterQuota(QuotaBuilder) error
+	RegisterQuota(QuotaBuilder)
 }
 
-// RegisterFn is a function the mixer invokes to trigger adapters to registers
-// their aspect builders.
-type RegisterFn func(Registrar) error
+// MustRegisterFn is a function the mixer invokes to trigger adapters to register
+// their aspect builders. It must succeed or panic().
+type MustRegisterFn func(Registrar)
```
Comment (mandarjog) [inline pkg/adapter/registrar.go]: I have re-named it back to `Register`. There are prior examples of `MustRegister` which means `RegisterOrPanic`

File: pkg/config/manager.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+	lastError error
+}
+
+// ManagerArgs are constructor args for a manager
+type ManagerArgs struct {
```
Comment (ZackButcher) [inline pkg/config/manager.go]: Same as before, I think the struct-as-args is an anti-pattern. Where I've seen struct-as-args used the struct represents options that affect the behavior of the thing being constructed, and they're typically optional. In this case all the arguments are required (except loop delay, I suppose), and you even embed the Args struct directly into the manager. You also don't get any encapsulation since all of these fields are public to enable construction outside of the package; since the Args struct is embedded in the Manager I can change the values arbitrarily after construction. Even though 6 args to the NewManager func isn't great, I think it's better than the struct version.

Alternatively, do what [gRPC does, have a private options struct](https://github.com/grpc/grpc-go/blob/master/server.go#L107) (or, in this case, args struct) and provide functions to operate on them. But, since all of these args are required that just forces more boilerplate everywhere.

File: pkg/config/manager.go [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+	cl      []ChangeListener
+	closing chan bool
+	scSHA   [sha1.Size]byte
+	gcSHA   [sha1.Size]byte
+
+	sync.RWMutex
+	lastError error
+}
+
+// NewManager returns a config.Manager given ManagerArgs.
```
Comment (geeknoid) [inline pkg/config/manager.go]: Comment is out of date.

File: pkg/api/handler.go [hunk 10]
```diff
@@ -119,10 +88,24 @@ func (h *methodHandlers) execute(ctx context.Context, tracker attribute.Tracker,
 
 	// get a new context with the attribute bag attached
 	ctx = attribute.NewContext(ctx, ab)
-	for _, conf := range h.configs[method] {
+
+	cfg := h.cfg.Load().(*config.Runtime)
+
+	if cfg == nil {
+		// config has NOT been loaded yet
+		glog.Warningf("configuration is not available")
```
Comment (mandarjog) [inline pkg/api/handler.go]: done.

**Comment by mandarjog** [comment]
Addressed review comments.  PTAL.

File: adapter/inventory.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+import (
+	"istio.io/mixer/adapter/denyChecker"
+	"istio.io/mixer/adapter/genericListChecker"
+	"istio.io/mixer/adapter/ipListChecker"
+	"istio.io/mixer/adapter/stdioLogger"
+	"istio.io/mixer/pkg/adapter"
+)
+
+// Inventory  returns inventory of all available adapters.
```
Comment (douglas-reid) [inline adapter/inventory.go]: nit: extra space between Inventory and returns.

File: cmd/server/server.go [hunk 12]
```diff
@@ -26,15 +26,17 @@ import (
...
+	"istio.io/mixer/pkg/adapterManager"
 	"istio.io/mixer/pkg/api"
+	"istio.io/mixer/pkg/aspect"
 	"istio.io/mixer/pkg/attribute"
+	"istio.io/mixer/pkg/config"
 	"istio.io/mixer/pkg/tracing"
 
-	denyadapter "istio.io/mixer/adapter/denyChecker"
-	"istio.io/mixer/pkg/aspect"
+	"time"
```
Comment (douglas-reid) [inline cmd/server/server.go]: this looks like it needs cleanup... i would expect std packages to be grouped, etc.

File: pkg/config/manager_test.go [hunk 13]
```diff
@@ -0,0 +1,141 @@
+package config
```
Comment (geeknoid) [inline pkg/config/manager_test.go]: Need copyright.

File: pkg/aspect/accessLoggerManager_test.go [hunk 14]
```diff
@@ -19,14 +19,16 @@ import (
 	"testing"
 	"text/template"
 
-	"github.com/golang/protobuf/ptypes/struct"
 	"istio.io/mixer/pkg/adapter"
-	"istio.io/mixer/pkg/aspect/config"
+	aconfig "istio.io/mixer/pkg/aspect/config"
```
Comment (douglas-reid) [inline pkg/aspect/accessLoggerManager_test.go]: looks like we could clean this up a bit.  awkward newline between groups of external imports... if we want to keep that, i'd suggest moving the named imports to their own block and otherwise combining.

File: pkg/aspect/loggerManager_test.go [hunk 15]
```diff
@@ -19,17 +19,38 @@ import (
...
 	"istio.io/mixer/pkg/adapter"
-	"istio.io/mixer/pkg/aspect/config"
 	"istio.io/mixer/pkg/aspect/test"
+
+	aconfig "istio.io/mixer/pkg/aspect/config"
 	"istio.io/mixer/pkg/attribute"
 	"istio.io/mixer/pkg/expr"
 
-	configpb "istio.io/api/mixer/v1/config"
+	"errors"
```
Comment (douglas-reid) [inline pkg/aspect/loggerManager_test.go]: as before, this should be cleaned up.

File: pkg/aspect/loggerManager_test.go [hunk 16]
```diff
@@ -19,17 +19,38 @@ import (
...
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/empty"
 	dpb "istio.io/api/mixer/v1/config/descriptor"
+	"istio.io/mixer/pkg/config"
+	configpb "istio.io/mixer/pkg/config/proto"
+)
+
+type (
```
Comment (douglas-reid) [inline pkg/aspect/loggerManager_test.go]: this stuff was killed in a recent CL. Let's not bring it back. In particular, testLogger has been moved to test.Logger.

File: pkg/aspect/loggerManager_test.go [hunk 17]
```diff
@@ -254,3 +268,20 @@ func TestLoggerManager_ValidateConfig(t *testing.T) {
 		t.Errorf("ValidateConfig(): unexpected error: %v", err)
 	}
 }
+
+func (t *testLogger) NewLogger(e adapter.Env, m adapter.AspectConfig) (adapter.LoggerAspect, error) {
```
Comment (douglas-reid) [inline pkg/aspect/loggerManager_test.go]: this stuff is all available in the test package.  not needed here.

File: pkg/aspect/manager.go [hunk 18]
```diff
@@ -20,19 +20,16 @@ package aspect
 import (
 	"google.golang.org/genproto/googleapis/rpc/code"
 
-	istioconfig "istio.io/api/mixer/v1/config"
+	"istio.io/mixer/pkg/config"
+
+	"io"
```
Comment (douglas-reid) [inline pkg/aspect/manager.go]: clean.

File: pkg/config/manager.go [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"crypto/sha1"
+	"io/ioutil"
+	"time"
+
```
Comment (douglas-reid) [inline pkg/config/manager.go]: clean.

**Comment by douglas-reid** [comment]
Can we edit the description for the PR to provide more context and a fuller description of the changes?  There is a lot going on here, and it would be nice to document some of the reasoning behind and capabilities provided by this PR.

**Comment by mandarjog** [comment]
I will update the description and update the test package changes.

For package re ordering can that be automated ? 
Do you use a tool?

Regards, 
Sent from my Motorola Next

Mandar U Jog

File: pkg/config/manager_test.go (hunk 13 above)
Comment (mandarjog) [inline pkg/config/manager_test.go]: done.

**Comment by mandarjog** [comment]
Review changes made.

re: goimports, I have only made changes that can be automated.
I don't think we should hand craft an ordering. 
I have used `goimports -local istio.io  -w` 
The order being "stardard imports, third party imports, istio imports"

I have also added check_license to linters.

File: pkg/config/manager_test.go [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+import (
+	"io/ioutil"
+	"os"
+	"strings"
+	"testing"
+
+	"istio.io/mixer/pkg/adapter"
+
+	"sync"
+	"time"
```
Comment (douglas-reid) [inline pkg/config/manager_test.go]: this still doesn't look correct to me. I know that you've run goimports, but there is an obvious flaw here. "sync" and "time" should be moved up with the other standard import packages and the two "istio.io/mixer/..." blocks should be combined into one.

File: pkg/adapterManager/manager.go [hunk 21]
```diff
@@ -18,122 +18,193 @@ import (
 	"fmt"
 	"sync"
```
Comment (douglas-reid) [inline pkg/adapterManager/manager.go]: any idea why there is an extra newline here? it feels like the "bytes"... section should be combined (and sorted) with the "fmt" and "sync" import block.

File: pkg/config/manager_test.go (hunk 20 above)
Comment (mandarjog) [inline pkg/config/manager_test.go]: done.
I need to remove blank lines either manually, or thru gogland "Code cleanup".
gofmt / goimports does not eliminate blank lines.

File: bin/check_license.sh [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+	if [[ $fn == *.pb.go ]];then
+		continue
+	fi
+	head -20 $fn | grep "Apache License, Version 2" > /dev/null
+	if [[ $? -ne 0 ]]; then
+		echo "${fn} missing license"
+		ret=$(($ret+1))
+	fi
+
+	head -20 $fn | grep Copyright | grep Google > /dev/null
```
Comment (geeknoid) [inline bin/check_license.sh]: Can't require a Google copyright in an OSS project, right?

File: bin/linters.sh [hunk 23]
```diff
@@ -64,4 +64,6 @@ prep_linters
 
 run_linters
 
+$SCRIPTPATH/check_license.sh
```
Comment (geeknoid) [inline bin/linters.sh]: This should be moved up to be in the run_linters function.

**Comment by mandarjog** [comment]
Done.
